diffkemp/semdiff/function_diff.py

```diff
@@ -124,9 +126,117 @@ def functions_semdiff(first, second, fun_first, fun_second, config):
+                      for call in callstack])
+
+
+def graph_to_fun_pair_list(fun_first, fun_second, graph):
+    # Extract the functions that should be compared from the graph in
+    # the form of Vertex objects.
+    vertices_to_compare = graph.reachable_from(
+        ComparisonGraph.Side.LEFT, fun_first)
+    # Use methods from ComparisonGraph (on the graph variable) and
+    # vertices_to_compare to generate objects_to_compare.
+    objects_to_compare = []
+    syndiff_bodies_left = dict()
+    syndiff_bodies_right = dict()
+    for vertex in vertices_to_compare:
+        if vertex.result in [Result.Kind.EQUAL,
+                             Result.Kind.ASSUMED_EQUAL]:
+            # Do not include equal functions into the result.
+            continue
+        # Generate and add the function difference.
+        fun_pair = []
+        for side in [ComparisonGraph.Side.LEFT,
+                     ComparisonGraph.Side.RIGHT]:
+            fun = fun_first if side == ComparisonGraph.Side.LEFT \
+                else fun_second
+            if fun == vertex.name[side]:
+                # There is no callstack from the base function.
+                calls = None
+            else:
+                # Transform the Edge objects returned by
+                # get_shortest_path to a readable callstack.
+                try:
+                    edges = graph.get_callstack(
+                        side, fun, vertex.name[side])
+                except ValueError:
+                    raise SimpLLException()
+                calls = _edge_callstack_to_string(edges)
+            # Note: a function diff is covered (i.e. hidden when empty
+            # if and only if there is a non-function difference
+            # referencing it).
+            fun_pair.append(Result.Entity(
+                vertex.name[side],
+                vertex.file[side],
+                vertex.line[side],
+                calls,
+                "function",
+                len(vertex.nonfun_diffs) != 0
+            ))
+        fun_pair.append(vertex.result)
+        objects_to_compare.append(tuple(fun_pair))
+
+        # Process non-function differences.
+        for nonfun_diff in vertex.nonfun_diffs:
+            nonfun_pair = []
+            for side in [ComparisonGraph.Side.LEFT,
+                         ComparisonGraph.Side.RIGHT]:
+                syndiff_bodies = (syndiff_bodies_left
+                                  if side == ComparisonGraph.Side.LEFT
+                                  else syndiff_bodies_right)
+                # Convert the YAML callstack format to string.
+                calls = ["{} at {}:{}".format(call["function"],
+                                              call["file"],
+                                              call["line"])
+                         for call in nonfun_diff.callstack[side]]
+                # Join the elements in the list to get a string.
+                calls = "\n".join(calls)
```

Comment:
This reminds body of the `_edge_callstack_to_string` function. Could it be somehow merged?

Reply:
It could, but that would require having a function take two different types of values as its parameter, since a dictionary is used in one case and a class in the other. (And the use of a trick like renaming the YAML attributes and passsing `call.__dict__` in the other case seems ugly to me.)

Reply:
Right, in that case, at least enclose this to a function, too. I.e., we'll have two functions with same functionality for two different types.